// wdio-vscode-service/dist/pageobjects/menu/WindowControls.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import logger from '@wdio/logger';
import { PageDecorator, BasePage } from '../utils.js';
import { TitleBar } from '../../index.js';
import { WindowControls as WindowControlsLocators } from '../../locators/1.73.0.js';
const log = logger('wdio-vscode-service');
/**
 * Page object for the windows controls part of the title bar
 *
 * @category Menu
 */
let WindowControls = class WindowControls extends BasePage {
    constructor(locators, element, bar = new TitleBar(locators)) {
        super(locators, element);
        this.bar = bar;
        /**
         * @private
         */
        this.locatorKey = 'WindowControls';
        this.setParentElement(this.bar.elem);
    }
    /**
     * Use the minimize window button
     * @returns Promise resolving when minimize button is pressed
     */
    async minimize() {
        await this.minimize$.click();
    }
    /**
     * Use the maximize window button if the window is not maximized
     * @returns Promise resolving when maximize button is pressed
     */
    async maximize() {
        try {
            await this.maximize$.click();
        }
        catch (err) {
            log.error('Window is already maximized', err);
        }
    }
    /**
     * Use the restore window button if the window is maximized
     * @returns Promise resolving when restore button is pressed
     */
    async restore() {
        try {
            await this.restore$.click();
        }
        catch (err) {
            log.error('Window is not maximized', err);
        }
    }
    /**
     * Use the window close button. Use at your own risk.
     * @returns Promise resolving when close button is pressed
     */
    async close() {
        await this.close$.click();
    }
};
WindowControls = __decorate([
    PageDecorator(WindowControlsLocators)
], WindowControls);
export { WindowControls };
