// wdio-vscode-service/dist/utils.js
import fs from 'node:fs/promises';
import url from 'node:url';
import path from 'node:path';
import { VSCODE_CAPABILITY_KEY } from './constants.js';
const __dirname = url.fileURLToPath(new URL('.', import.meta.url));
export async function getLocators(version) {
    if (version === 'insiders') {
        return import('./locators/insiders.js');
    }
    const files = (await fs.readdir(path.join(__dirname, 'locators'), { encoding: 'utf-8' }))
        .filter((filename) => filename.endsWith('.js') && !filename.endsWith('.d.ts'))
        .map((filename) => filename.slice(0, -3));
    const [major, minor] = version.split('.');
    const sanitizedVersion = `${major}.${minor}.0`;
    const locatorFile = files.find((f, i) => (f === sanitizedVersion
        || (files[i + 1] && files[i + 1] > sanitizedVersion))) || files[files.length - 1];
    return import(`./locators/${locatorFile}.js`);
}
export function fileExist(filepath) {
    return fs.access(filepath).then(() => true, () => false);
}
export async function directoryExists(directoryPath) {
    try {
        const stats = await fs.stat(directoryPath);
        return stats.isDirectory();
    }
    catch {
        return false;
    }
}
export function getValueSuffix(value) {
    if (typeof value === 'boolean' && value) {
        return '';
    }
    return `=${value}`;
}
export function isVSCodeCapability(cap) {
    return Boolean(cap[VSCODE_CAPABILITY_KEY]);
}
var FileType;
(function (FileType) {
    FileType[FileType["Unknown"] = 0] = "Unknown";
    FileType[FileType["File"] = 1] = "File";
    FileType[FileType["Directory"] = 2] = "Directory";
    FileType[FileType["SymbolicLink"] = 64] = "SymbolicLink";
})(FileType || (FileType = {}));
export function getFileType(stats) {
    if (stats.isFile()) {
        return FileType.File;
    }
    if (stats.isDirectory()) {
        return FileType.Directory;
    }
    if (stats.isSymbolicLink()) {
        return FileType.SymbolicLink;
    }
    return FileType.Unknown;
}
