// wdio-electron-service/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channel = exports.CUSTOM_CAPABILITY_NAME = exports.APP_NAME_DETECTION_ERROR = exports.BUILD_TOOL_DETECTION_ERROR = exports.MULTIPLE_BUILD_TOOLS_ERROR = exports.APP_NOT_FOUND_ERROR = void 0;
exports.APP_NOT_FOUND_ERROR = 'Could not find Electron app at %s built with %s!\n' +
    'If the application is not compiled, please do so before running your tests, e.g. via `%s`.\n' +
    'Otherwise if the application is compiled at a different location, please specify the `appBinaryPath` option in your capabilities.';
exports.MULTIPLE_BUILD_TOOLS_ERROR = 'Multiple build tools were detected, please remove configuration and dependencies for tools which are not being used to build your application.';
exports.BUILD_TOOL_DETECTION_ERROR = 'No build tool was detected, if the application is compiled at a different location, please specify the `appBinaryPath` option in your capabilities.';
exports.APP_NAME_DETECTION_ERROR = 'No application name was detected, please set name / productName in your package.json or build tool configuration.';
exports.CUSTOM_CAPABILITY_NAME = 'wdio:electronServiceOptions';
var Channel;
(function (Channel) {
    Channel["Execute"] = "wdio-electron.execute";
})(Channel || (exports.Channel = Channel = {}));
